### PR TITLE
Output diffs on mix format --check-formatted

### PR DESCRIPTION
**Pull request by NickNeck**
A suggestion for #12100 . The output would look like:
diff
In case a single line was changed the changed spaces get a background colour. For multiple deleted or inserted lines just the text gets a colour.

I have written this for another project but maybe it fits in here.

### CODE REVIEW

File: lib/mix/lib/mix/text_diff.ex [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+  @spec format(String.t(), String.t(), keyword()) :: iodata()
+  def format(code, code, opts \\ default_opts())
+
+  def format(code, code, _opts), do: []
+
+  def format(old, new, opts) do
+    opts = Keyword.merge(@default_opts, opts)
+
+    old = String.split(old, "\n")
+    new = String.split(new, "\n")
```
Comment (josevalim) [inline lib/mix/lib/mix/text_diff.ex:136]:
```suggestion
    old = String.split(old, ["\r\n", "\n"])
    new = String.split(new, ["\r\n", "\n"])
```

**Comment by josevalim** [comment]
@NickNeck holy moly, this is awesome! :heart:

I think we want to keep the `Mix.TextDiff` module private though, unless @whatyouhide @sabiwara @ericmj @fertapric have a good suggestion of where it should be placed in the Elixir API.

File: lib/mix/lib/mix/tasks/format.ex [hunk 2]
```diff
@@ -651,11 +651,23 @@ defmodule Mix.Tasks.Format do
...
+  defp to_diffs(files) do
+    Enum.map_join(files, "\n", fn
+      {:stdin, unfomatted, fomatted} ->
+        IO.iodata_to_binary([IO.ANSI.reset(), Mix.TextDiff.format(unfomatted, fomatted)])
+
+      {file, unfomatted, fomatted} ->
+        IO.iodata_to_binary([
+          IO.ANSI.red(),
+          file,
+          IO.ANSI.reset(),
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/format.ex:667]:
```suggestion
          file,
          "\n",
          IO.ANSI.reset(),
```

File: lib/mix/lib/mix/text_diff.ex (hunk 1 above)
Comment (josevalim) [reply lib/mix/lib/mix/text_diff.ex:136]: But then what happens when two files are equal except for `\r\n` -> `\n`? So maybe it is better to keep it by your current split?

**Comment by josevalim** [comment]
@NickNeck let's make `Mix.TextDiff` a function with `@doc false` at the bottom of `Mix.Tasks.Format` for now, so we can unit test it, and later on we will figure out a public API for it.

**Comment by sabiwara** [comment]
> @NickNeck holy moly, this is awesome! ❤️
> 
> I think we want to keep the `Mix.TextDiff` module private though, unless @whatyouhide @sabiwara @ericmj @fertapric have a good suggestion of where it should be placed in the Elixir API.

I'm not sure, but maybe in the future it might be interesting to have a public `Diff` / `Inspect.Diff` / `Debug.Diff` module in Elixir exposing what is currently done by `ExUnit.Diff` and this new text diff functionality? But it would be good to have more use case examples pointing to the fact these could be reused out of their original context (ExUnit, Mix).

File: lib/mix/lib/mix/tasks/format.ex [hunk 3]
```diff
@@ -651,11 +651,23 @@ defmodule Mix.Tasks.Format do
...
+    #{to_diffs(not_formatted)}
     """)
   end
 
-  defp to_bullet_list(files) do
-    Enum.map_join(files, "\n", &"  * #{&1 |> to_string() |> Path.relative_to_cwd()}")
+  defp to_diffs(files) do
+    Enum.map_join(files, "\n", fn
+      {:stdin, unfomatted, fomatted} ->
+        IO.iodata_to_binary([IO.ANSI.reset(), Mix.TextDiff.format(unfomatted, fomatted)])
```
Comment (sabiwara) [inline lib/mix/lib/mix/tasks/format.ex:661]: I'm probably overthinking this, but just wondering: maybe we don't need to cast it as a binary here and we can build an IO list?
- removing calls to `IO.iodata_to_binary/1`
- replacing `map_join/3` by `map_intersperse/3`

File: lib/mix/lib/mix/text_diff.ex (hunk 1 above)
Comment (NickNeck) [reply lib/mix/lib/mix/text_diff.ex:139]: Yes, splitting with `\n` should be better. The last commit makes also changed `\r` visible.
```
iex(1)> Mix.TextDiff.format("a\r\nb\r\n", "a\nb\nc\n") |> IO.iodata_to_binary() |> IO.puts()
1   - |a↵
2   - |b↵
  1 + |a
  2 + |b
  3 + |c

:ok
iex(2)> Mix.TextDiff.format("a\r\n", "a\n") |> IO.iodata_to_binary() |> IO.puts()
1   - |a↵
  1 + |a

:ok
iex(3)> Mix.TextDiff.format("a\n", "a\r\n") |> IO.iodata_to_binary() |> IO.puts()
1   - |a
  1 + |a↵

:ok
```

**Comment by josevalim** [comment]
@NickNeck looks great! We just need to move TextDiff around and we can ship it! 👍

File: lib/mix/lib/mix/tasks/format.ex [hunk 4]
```diff
@@ -651,11 +651,24 @@ defmodule Mix.Tasks.Format do
...
-    #{to_bullet_list(not_formatted)}
+    #{to_diffs(not_formatted)}
     """)
   end
 
-  defp to_bullet_list(files) do
-    Enum.map_join(files, "\n", &"  * #{&1 |> to_string() |> Path.relative_to_cwd()}")
+  defp to_diffs(files) do
+    Enum.map_intersperse(files, "\n", fn
+      {:stdin, unfomatted, fomatted} ->
```
Comment (davydog187) [inline lib/mix/lib/mix/tasks/format.ex:660]: I think you're missing an "r" on these ☺️

**Comment by NickNeck** [comment]
> @NickNeck let's make Mix.TextDiff a function with @doc false at the bottom of Mix.Tasks.Format for now, so we can unit test it, and later on we will figure out a public API for it.

@josevalim I have moved `Mix.TextDiff` to `Mix.Tasks.Format.text_diff_format/3`.

**Comment by josevalim** [comment]
@NickNeck looks great! Can you please look at the failing tests?

Actually, I would suggest calling `IO.iodata_to_binary/1` on the output of the function, as that would make those tests cleaner and resilient to refactoring. :)

**Pull request merged**